src/vsr.zig

```diff
@@ -336,6 +371,141 @@ pub const Header = packed struct {
         return null;
     }
 
+    fn invalid_reply(self: *const Header) ?[]const u8 {
+        assert(self.command == .reply);
+        if (self.client == 0) return "client == 0";
+        if (self.context != 0) return "context != 0";
+        if (self.op != self.commit) return "reserved: op != commit";
+        if (self.operation != .register) {
+            if (self.commit == 0) return "commit == 0";
+            if (self.request == 0) return "request == 0";
+        }
+        return null;
+    }
+
+    fn invalid_commit(self: *const Header) ?[]const u8 {
+        assert(self.command == .commit);
+        if (self.parent != 0) return "parent != 0";
+        if (self.client != 0) return "client != 0";
+        if (self.context == 0) return "context == 0";
+        if (self.request != 0) return "request != 0";
+        if (self.op != 0) return "op != 0";
+        if (self.offset != 0) return "offset != 0";
+        if (self.operation != .reserved) return "operation != .reserved";
+        return null;
+    }
+
+    fn invalid_start_view_change(self: *const Header) ?[]const u8 {
+        assert(self.command == .start_view_change);
+        if (self.parent != 0) return "parent != 0";
+        if (self.client != 0) return "client != 0";
+        if (self.context != 0) return "context != 0";
+        if (self.request != 0) return "request != 0";
+        if (self.op != 0) return "op != 0";
+        if (self.commit != 0) return "commit != 0";
+        if (self.offset != 0) return "offset != 0";
+        if (self.operation != .reserved) return "operation != .reserved";
+        return null;
+    }
+
+    fn invalid_do_view_change(self: *const Header) ?[]const u8 {
+        assert(self.command == .do_view_change or self.command == .start_view);
+        if (self.parent != 0) return "parent != 0";
+        if (self.client != 0) return "client != 0";
+        if (self.context != 0) return "context != 0";
+        if (self.request != 0) return "request != 0";
+        if (self.operation != .reserved) return "operation != .reserved";
+        return null;
+    }
+
+    fn invalid_recovery(self: *const Header) ?[]const u8 {
+        assert(self.command == .recovery);
+        // TODO implement validation once the message is actually used.
+        return null;
+    }
+
+    fn invalid_recovery_response(self: *const Header) ?[]const u8 {
+        assert(self.command == .recovery_response);
+        if (self.parent != 0) return "parent != 0";
+        if (self.client != 0) return "client != 0";
+        if (self.request != 0) return "request != 0";
+        if (self.offset != 0) return "offset != 0";
+        if (self.op == 0) return "op == 0";
+        if (self.operation != .reserved) return "operation != .reserved";
```

Comment:
```suggestion
        // TODO implement validation once the message is actually used.
```

We may remove these recovery handlers, switching to persisting the view number to disk (with a few redundant copies on disk) before we ack a view transition externally as part of the view change protocol.

The rationale:

1. There's a correctness bug in the 2012 VSR paper for the recovery protocol, cf. https://syslab.cs.washington.edu/papers/diskless-tr16.pdf. However, the solution in there, if I understand correctly, seems to require unbounded vector clocks, which might be a little too complex. I am trying to get hold of the patched VSR implementation that the paper evaluated — it would be nice to see the impact of the fix on the actual recovery code.
2. The patched recovery protocol would still need a further protocol to break replicas out of recovery if all replicas were power-cycled at the same time, provided there is consensus on the cluster state.
3. Whereas, if we persist view numbers, then we're doing the 1988 Brian Oki flavor view change, so we can drop the round of `start_view_change` messages. The 2012 determinism plus the Brian Oki flavor is then pretty optimal: no dueling leaders, and less messages in the network, therefore less chance for network weather to impede the view change.

Reply:
Also just to clarify that if we do persist view numbers to disk, then there's no longer any need for the recovery protocol.

If we want to be able to handle all copies of on-disk metadata getting corrupted simultaneously, then we can explore either the reconfiguration protocol or recovery protocol, but that's further down the road.